sw/otbn/code-snippets/calc_montgomery_consts.s

```diff
@@ -0,0 +1,285 @@
+/* Copyright lowRISC contributors. */
+/* Licensed under the Apache License, Version 2.0, see LICENSE for details. */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+.text
+/**
+ * Compute the inverse of an odd number modulo 2^{256}.
+ *
+ *   Returns -m^{-1} mod 2^{256} given odd m
+ *
+ * This routine runs in constant time.
+ *
+ * @param[in]:   w0:     odd 256-bit number m
+ * @param[out]:  w0:     m0' = -m^{-1} mod 2^{256}
+ * clobbered registers:    w1-w2
+ * clobbered flag groups:  FG0
+ * called subroutines:     -
+ */
+ .section .text.compute_m0inv2to256
+ .balign 4
+ .global compute_m0inv2to256
+ .type compute_m0inv2to256, @function
+
+compute_m0inv2to256:
+  /* As odd numbers are their own inverse mod 8, the input is taken as
+      start value for a Newton iteration (quadratic Hensel lifting)
+      doubling the precision in each step.
+     The code can be adapted to calculate -m^-1 mod 3*2^k by
+      increasing the precision of the last round of first loop to
+      96 bit, and of the following loops to 192 rsp. 384 bit. */
+  bn.mov          w1, w0  /* w0 is its own inverse with 3-bit precision */
+  /* double precision via Newton in each round from 2^1 to 2^6 bit */
+  loopi           5, 4
+    bn.mulqacc.wo.z w2, w1.0, w0.0, 0
+    bn.mulqacc.wo.z w2, w1.0, w2.0, 0
+    bn.add          w1, w1, w1
+    bn.sub          w1, w1, w2
+  /* double precision from 64 to 128 bit */
+  bn.mov          w2, w0
+  loopi           2, 3
+    bn.mulqacc.z    w1.0, w2.0, 0
+    bn.mulqacc      w1.0, w2.1, 64
+    bn.mulqacc.wo   w2, w1.1, w2.0, 64
+  bn.add          w1, w1, w1
+  bn.sub          w1, w1, w2
+  /* double precision from 128 to 256 bit */
+  loopi           2, 10
+    bn.mulqacc.z    w1.0, w0.0, 0
+    bn.mulqacc      w1.0, w0.1, 64
+    bn.mulqacc      w1.1, w0.0, 64
+    bn.mulqacc      w1.0, w0.2, 128
+    bn.mulqacc      w1.1, w0.1, 128
+    bn.mulqacc      w1.2, w0.0, 128
+    bn.mulqacc      w1.0, w0.3, 192
+    bn.mulqacc      w1.1, w0.2, 192
+    bn.mulqacc      w1.2, w0.1, 192
+    bn.mulqacc.wo   w0, w1.3, w0.0, 192
+  bn.add          w1, w1, w1
+  bn.sub          w0, w0, w1
+  ret
+ .size compute_m0inv2to256, .-compute_m0inv2to256
+
+/* Calculate r^2 mod m, where r = 2^{256 k} and 2^{256*k-1} < m < r.
+ *
+ *   Returns 4^{256*k} mod m given m with 2^{256*k-1} < m < 2^{256*k}
+ *
+ * This routine runs in constant time (for fixed k).
+ *
+ * @param[in]:   x10:    pointer to m, the most significant limb of m must be >=2^255
+ * @param[in]:   x11:    length k of m in 256-bit-limbs
+ * @param[in]:   x12:    pointer to output buffer for r^2 mod m, same size as m
+ *
+ * @param[out]:  x10:    0 if everything went ok, otherwise 1
+ *
+ * clobbered registers:    w0-w[3+k-1], x8-x10
+ * clobbered flag groups:  FG0, FG1
+ * called subroutines:     -
+ */
+  /* TODO: Montgomery squaring might be faster at some k, as soon r*2^k is known. */
```

Comment:
I wanted to keep the sample code self-contained, so I chose not to include the optimization of Montgomery squaring in the code snippet. The idea would rather be to replace lines 168-184 of rsa_verify_3072_rr.s by an adaption of the sample code (correct registers; outer loop runs only 96 instead of 1536 rounds). The shorter sample code should be about 30% faster than the lines to be replaced, and the faster version about another factor of 2 (probably it's better to run it 192 rounds and Montgomery-square only 4 times). If constant time is not needed, the subroutine called in the inner loop of the faster code can be split as described in its comment to gain another 30%. For the case of 3072 bit specializing the doubling mod m to fixed width probably gains less than 10%.

Reply:
(A comment about calling montmul in rsa_verify.s from compute_rr in rsa_verify_3072_rr.s: I couldn't find where register x30 is set to tell montmul the number of limbs.)